src/main/kotlin/com/lykke/client/accounts/ClientAccountsCacheImpl.kt

```diff
@@ -2,16 +2,33 @@ package com.lykke.client.accounts
 
 import java.util.concurrent.ConcurrentHashMap
 
-class ClientAccountsCacheImpl(private val clientIdByWalletId: ConcurrentHashMap<String, String>): ClientAccountsCache {
+class ClientAccountsCacheImpl(private val clientIdByWalletId: ConcurrentHashMap<String, String>) : ClientAccountsCache {
+    private val walletsByClientId = ConcurrentHashMap<String, MutableSet<String>>()
+
+    init {
+        clientIdByWalletId.forEach { walletId, clientId ->
+            val wallets = walletsByClientId.getOrPut(clientId) { ConcurrentHashMap.newKeySet() }
+            wallets.add(walletId)
+        }
+    }
+
     internal fun delete(walletId: String) {
-        clientIdByWalletId.remove(walletId)
+        val clientId = clientIdByWalletId.remove(walletId) ?: return
+        val wallets = walletsByClientId[clientId] ?: return
+        wallets.remove(walletId)
     }
 
     internal fun add(clientId: String, walletId: String) {
         clientIdByWalletId[walletId] = clientId
+        val wallets = walletsByClientId.getOrPut(clientId) { ConcurrentHashMap.newKeySet() }
```

Comment:
This can be extracted to a separate method and reused in `init` block

Reply:
fixed